### PR TITLE
Added test around getting latest arrears agreements by start date

### PR DESCRIPTION
**Pull request by StevenLeighton21**
- Added new field to `arag` - `start_date` (not checked against real UH schema)
- Added  ArrearsAgreementStartDate to Domain  class.
- Added test around to confirm the latest start date is returned.

### CODE REVIEW

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs [hunk 1]
```diff
@@ -36,7 +37,7 @@ public List<TenancyListItem> GetTenanciesByRefs(List<string> tenancyRefs)
                                              $"ON arag.tag_ref = tenagree.tag_ref " +
                                              $"LEFT JOIN contacts " +
                                              $"ON contacts.tag_ref = tenagree.tag_ref " +
-                                             $"WHERE tenagree.tag_ref IN ('{tenancyRefs.Join("', '")}')")
+                                             $"WHERE tenagree.tag_ref IN ('{tenancyRefs.Join("', '")}') ORDER BY arag.start_date DESC")
```
Comment (zuren) [inline LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: Suggest the `ORDER BY` declaration goes on a new line for consistency.

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 2]
```diff
@@ -41,6 +41,21 @@ public void WhenGivenTenancyRef_GetTenanciesByRefs_ShouldReturnTenancyObjectForT
...
 
+        [Fact]
+        public void WhenGivenTenancyRef_GetTenanciesByRefs_ShouldReturnTheLatestAgreement()
+        {
+            TenancyListItem expectedTenancy = CreateRandomTenancyListItem();
+            InsertTenancyAttributes(expectedTenancy);
+
+            DateTime latestAragDate = expectedTenancy.ArrearsAgreementStartDate.AddDays(1);
+            InsertAgreement(expectedTenancy.TenancyRef, "Inactive", expectedTenancy.ArrearsAgreementStartDate.Subtract(new TimeSpan(1, 0, 0, 0)));
+            InsertAgreement(expectedTenancy.TenancyRef, "Active", latestAragDate);
```
Comment (zuren) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Would it be worth using named arguments for readability?

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 3]
```diff
@@ -93,5 +112,20 @@ private void InsertTenancyAttributes(TenancyListItem tenancyAttributes)
...
+            SqlCommand command = new SqlCommand(commandText, db);
+            command.Parameters.Add("@tenancyRef", SqlDbType.NVarChar);
+            command.Parameters["@tenancyRef"].Value = tagRef;
+            command.Parameters.Add("@agreementStatus", SqlDbType.NVarChar);
+            command.Parameters["@agreementStatus"].Value = status;
+            command.Parameters.Add("@startDate", SqlDbType.SmallDateTime);
+            command.Parameters["@startDate"].Value = startDate;
+
+            command.ExecuteNonQuery();
+        }
```
Comment (zuren) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: There's now duplication of the SQL used to insert an agreement - suggest we modify `InsertTenancyAttributes` to remove the duplicated SQL and instead call this new function.

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs [hunk 4]
```diff
@@ -26,6 +26,7 @@ public List<TenancyListItem> GetTenanciesByRefs(List<string> tenancyRefs)
                                              $"araction.action_code as LastActionCode, " +
                                              $"araction.action_date as LastActionDate, " +
                                              $"arag.arag_status as ArrearsAgreementStatus, " +
+                                             $"arag.start_date as ArrearsAgreementStartDate, " +
```
Comment (zuren) [inline LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: Why did you choose not to check against the real schema?

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 5]
```diff
@@ -41,6 +41,21 @@ public void WhenGivenTenancyRef_GetTenanciesByRefs_ShouldReturnTenancyObjectForT
...
         }
 
+        [Fact]
+        public void WhenGivenTenancyRef_GetTenanciesByRefs_ShouldReturnTheLatestAgreement()
+        {
+            TenancyListItem expectedTenancy = CreateRandomTenancyListItem();
+            InsertTenancyAttributes(expectedTenancy);
+
+            DateTime latestAragDate = expectedTenancy.ArrearsAgreementStartDate.AddDays(1);
+            InsertAgreement(expectedTenancy.TenancyRef, "Inactive", expectedTenancy.ArrearsAgreementStartDate.Subtract(new TimeSpan(1, 0, 0, 0)));
```
Comment (zuren) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: `Subtract(new TimeSpan(1, 0, 0, 0))` is awkward but I suppose conventional? Would it be worth having a `day` var to make this more readable, or is it good enough as is? Could go either way tbh, drawbacks to either.

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs (hunk 2 above)
Comment (StevenLeighton21) [reply LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: image
That's why we name our args - named args in calls to functions make sense in dynamic languages but intellisense probably does enough for us in C# land

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs (hunk 3 above)
Comment (StevenLeighton21) [reply LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Didn't want to have to change other tests still using the function - do you think it's better to duplicate calls to `InsertAgreement` in each test or duplicate the SQL command in two places? I'm happy to change but it felt like the lesser of two DRY evils

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs (hunk 4 above)
Comment (StevenLeighton21) [reply LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: No VPN - I assume it'll have to be checked at some point very soon/we'll use the whole schema anyway

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs (hunk 3 above)
Comment (zuren) [reply LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Why not call `InsertAgreement` inside `InsertTenancyAttributes`? That way you don't have to amend existing tests, behaviour will be the same, but we'll have no repetition?

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs (hunk 2 above)
Comment (zuren) [reply LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Up to you. Been using them in my use case code, like not having to rely on IntelliSense to decipher personally. But cool 👍

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs (hunk 3 above)
Comment (StevenLeighton21) [reply LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Yeah that'd be smart....

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 6]
```diff
@@ -113,13 +114,13 @@ private void InsertTenancyAttributes(TenancyListItem tenancyAttributes)
...
 
-        private void InsertAgreement(string tagRef, string status, DateTime startDate)
+        private void InsertAgreement(string tenancyRef, string status, DateTime startDate)
         {
             string commandText = "INSERT INTO arag (tag_ref, arag_status, start_date) VALUES (@tenancyRef, @agreementStatus, @startDate)";
 
             SqlCommand command = new SqlCommand(commandText, db);
             command.Parameters.Add("@tenancyRef", SqlDbType.NVarChar);
-            command.Parameters["@tenancyRef"].Value = tagRef;
+            command.Parameters["@tenancyRef"].Value = tenancyRef;
```
Comment (zuren) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Nice

**Pull request merged**